File: PdfConverterJava/app/src/main/java/com/example/pdfconverterjava/data/repository/PdfRepository.java
package com.example.pdfconverterjava.data.repository;


import com.example.pdfconverterjava.data.model.CreateJobResponse;
import com.example.pdfconverterjava.data.model.JobStatusResponse;
import com.example.pdfconverterjava.data.model.ListJobsResponse;
import com.example.pdfconverterjava.data.remote.PdfApiService;
import com.example.pdfconverterjava.data.remote.RetrofitClient;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;

public class PdfRepository {

    private final PdfApiService apiService;

    public PdfRepository() {
        apiService = RetrofitClient.getApiService();
    }

    /**
     * Tek dosyalı job oluşturur.
     * Örnek:
     * - jpg_to_pdf
     * - pdf_to_word
     * - word_to_pdf
     */
    public Call<CreateJobResponse> createSingleFileJob(int userId, String jobType, File file) {
        RequestBody jobTypeBody =
                RequestBody.create(jobType, MediaType.parse("text/plain"));

        RequestBody userIdBody =
                RequestBody.create(String.valueOf(userId), MediaType.parse("text/plain"));

        RequestBody fileRequestBody =
                RequestBody.create(file, MediaType.parse("application/octet-stream"));

        MultipartBody.Part filePart =
                MultipartBody.Part.createFormData("file", file.getName(), fileRequestBody);

        return apiService.createSingleFileJob(jobTypeBody, userIdBody, filePart);
    }

    /**
     * Çoklu dosyalı job oluşturur.
     * Örnek:
     * - pdf_merge
     */
    public Call<CreateJobResponse> createMultiFileJob(int userId, String jobType, List<File> files) {
        RequestBody jobTypeBody =
                RequestBody.create(jobType, MediaType.parse("text/plain"));

        RequestBody userIdBody =
                RequestBody.create(String.valueOf(userId), MediaType.parse("text/plain"));

        List<MultipartBody.Part> fileParts = new ArrayList<>();

        for (File file : files) {
            RequestBody requestBody =
                    RequestBody.create(file, MediaType.parse("application/pdf"));

            MultipartBody.Part part =
                    MultipartBody.Part.createFormData("files[]", file.getName(), requestBody);

            fileParts.add(part);
        }

        return apiService.createMultiFileJob(jobTypeBody, userIdBody, fileParts);
    }

    /**
     * Job durumunu getirir.
     */
    public Call<JobStatusResponse> getJobStatus(int jobId) {
        return apiService.getJobStatus(jobId);
    }

    /**
     * Kullanıcının geçmiş job listesini getirir.
     */
    public Call<ListJobsResponse> listJobs(int userId) {
        return apiService.listJobs(userId);
    }
}

File: PdfConverterKotlin/app/src/main/java/com/example/pdfconverterkotlin/data/repository/PdfRepository.kt
package com.example.pdfconverterkotlin.data.repository

import com.example.pdfconverterkotlin.data.model.CreateJobResponse
import com.example.pdfconverterkotlin.data.model.JobStatusResponse
import com.example.pdfconverterkotlin.data.model.ListJobsResponse
import com.example.pdfconverterkotlin.data.remote.RetrofitClient
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File

class PdfRepository {

    private val api = RetrofitClient.api

    suspend fun createSingleFileJob(
        userId: Int,
        jobType: String,
        file: File
    ): Result<CreateJobResponse> {
        return try {
            val jobTypeBody = jobType.toRequestBody("text/plain".toMediaTypeOrNull())
            val userIdBody = userId.toString().toRequestBody("text/plain".toMediaTypeOrNull())

            // Dosyanın mime type'ını burada basit tuttuk.
            // İstersen sonra dosya uzantısına göre dinamik yaparız.
            val requestFile: RequestBody =
                file.asRequestBody("application/octet-stream".toMediaTypeOrNull())

            val filePart = MultipartBody.Part.createFormData(
                "file",
                file.name,
                requestFile
            )

            val response = api.createSingleFileJob(
                jobType = jobTypeBody,
                userId = userIdBody,
                file = filePart
            )

            if (response.isSuccessful && response.body() != null) {
                Result.success(response.body()!!)
            } else {
                Result.failure(Exception("createSingleFileJob başarısız: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun createMultiFileJob(
        userId: Int,
        jobType: String,
        files: List<File>
    ): Result<CreateJobResponse> {
        return try {
            val jobTypeBody = jobType.toRequestBody("text/plain".toMediaTypeOrNull())
            val userIdBody = userId.toString().toRequestBody("text/plain".toMediaTypeOrNull())

            val fileParts = files.map { file ->
                val requestBody = file.asRequestBody("application/pdf".toMediaTypeOrNull())

                MultipartBody.Part.createFormData(
                    "files[]",
                    file.name,
                    requestBody
                )
            }

            val response = api.createMultiFileJob(
                jobType = jobTypeBody,
                userId = userIdBody,
                files = fileParts
            )

            if (response.isSuccessful && response.body() != null) {
                Result.success(response.body()!!)
            } else {
                Result.failure(Exception("createMultiFileJob başarısız: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getJobStatus(jobId: Int): Result<JobStatusResponse> {
        return try {
            val response = api.getJobStatus(jobId)
            if (response.isSuccessful && response.body() != null) {
                Result.success(response.body()!!)
            } else {
                Result.failure(Exception("getJobStatus başarısız: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun listJobs(userId: Int): Result<ListJobsResponse> {
        return try {
            val response = api.listJobs(userId)
            if (response.isSuccessful && response.body() != null) {
                Result.success(response.body()!!)
            } else {
                Result.failure(Exception("listJobs başarısız: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}